packages/application-audit/src/middlewares.ts

```diff
@@ -0,0 +1,288 @@
+
+    if (!forwardedFor) {
+      throw genericInternalError("The forwardedFor header is missing");
+    }
+
+    const initialAudit: ApplicationAuditBeginRequest = {
+      correlationId,
+      service: serviceName,
+      serviceVersion: config.serviceVersion,
+      endpoint: req.path,
+      httpMethod: req.method,
+      phase: Phase.BEGIN_REQUEST,
+      requesterIpAddress: forwardedFor,
+      nodeIp: config.nodeIp,
+      podName: config.podName,
+      uptimeSeconds: Math.round(process.uptime()),
+      timestamp: requestTimestamp,
+      amazonTraceId: amznTraceId,
+    };
+
+    await producer.send({
+      messages: [
+        {
+          key: correlationId,
+          value: JSON.stringify(initialAudit),
+        },
+      ],
+    });
+
+    return next();
+  };
+}
+
+export async function applicationAuditEndMiddleware(
+  serviceName: string,
+  config: ApplicationAuditProducerConfig
+): Promise<RequestHandler> {
+  const producer = await initProducer(config, config.applicationAuditTopic);
+  return async (req, res, next): Promise<void> => {
+    if (
+      !config.endpointsWithCustomAudit ||
+      (config.endpointsWithCustomAudit &&
+        !config.endpointsWithCustomAudit.includes(req.route.path))
+    ) {
+      res.on("finish", async () => {
+        const context = (req as Request & { ctx?: AppContext }).ctx;
+        if (!context) {
+          throw genericInternalError("Failed to retrieve context");
+        }
+
+        const correlationId = context.correlationId;
+        const amznTraceId = parseAmznTraceIdHeader(req);
+        const forwardedFor = parseForwardedForHeader(req);
+
+        if (!forwardedFor) {
+          throw genericInternalError("The forwardedFor header is missing");
+        }
+
+        const endTimestamp = Date.now();
+
+        const finalAudit: ApplicationAuditEndRequest = {
+          correlationId,
+          service: serviceName,
+          serviceVersion: config.serviceVersion,
+          endpoint: req.route.path,
+          httpMethod: req.method,
+          phase: Phase.END_REQUEST,
+          requesterIpAddress: forwardedFor,
+          nodeIp: config.nodeIp,
+          podName: config.podName,
+          uptimeSeconds: Math.round(process.uptime()),
+          timestamp: endTimestamp,
+          amazonTraceId: amznTraceId,
+          organizationId: context.authData.organizationId,
+          userId: context.authData.userId,
+          httpResponseStatus: res.statusCode,
+          executionTimeMs: endTimestamp - context.requestTimestamp,
+        };
+
+        await producer.send({
+          messages: [
+            {
+              key: correlationId,
+              value: JSON.stringify(finalAudit),
+            },
+          ],
+        });
+      });
+    }
+
+    return next();
+  };
+}
+
+export async function applicationAuditEndSessionTokenExchangeMiddleware(
+  serviceName: string,
+  config: ApplicationAuditProducerConfig
+): Promise<RequestHandler> {
```

Comment:
This is called `SessionTokenExchangeMiddleware`, but it's actually generic - the endpoints are configurable, not hardcoded to `/session/tokens`; You should either hardcode the `/sessions/tokens` endpoint here or make the naming more generic. What is it that this middleware does differently from the normal `applicationAuditEndMiddleware`? We could give a meaningful name based on that.

Same rename to be applied to the audit type `ApplicationAuditEndRequestSessionTokenExchange`

Reply:
fields in the audit change, that's why the split

Reply:
Makes sense, the naming comment is still valid: if it's generic/configurable let's not tie the name to a specific endpoint